### PR TITLE
add Inverse wishart cholesky

### LINKED ISSUE #2714: add inverse wishart cholesky
Add implementations of Wishart density and RNG with Cholesky factor parameters and variates.  Signatures exposed through Stan will be

```stan
/**
 * Return normalized density of `L_Sigma` so that `L_Sigma * L_Sigma'`
 * has an `inverse_wishart(nu, L_Omega * L_Omega')` distribution.
 */
real inverse_wishart_cholesky_lpdf(matrix L_Sigma, real nu, matrix L_Omega);
```

```stan
/**
 * Return unnormalized density of `L_Sigma` so that `L_Sigma * L_Sigma'`
 * has an `inverse_wishart(nu, L_Omega * L_Omega')` distribution.
 */
real inverse_wishart_cholesky_lupdf(matrix L_Sigma, real nu, matrix L_Omega);
```

```stan
/**
 * Return Cholesky factor of a random inverse Wishart draw with degrees of
 * freedom `nu` and scale `L_Omega * L_Omega'`.
 */
matrix inverse_wishart_cholesky_rng(real nu, matrix L_Omega);
```

### PR DESCRIPTION
**Pull request by spinkney**
## Summary

Adds the `inv_wishart_cholesky_lpdf` and `_rng` functions. This PR depends upon #2710 for the `_rng` function. Closes #2714.

## Release notes

The inverse Wishart Cholesky parameterization `lpdf\lupdf` is available along with `inv_wishart_cholesky_rng`.

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Sean Pinkney

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  boost::random::mt19937 rng;
+
+  MatrixXd omega(3, 4);
+  EXPECT_THROW(inv_wishart_cholesky_rng(3.0, omega, rng), std::domain_error);
+
+  MatrixXd sigma(3, 3);
+  sigma << 9.0, -3.0, 0.0, -3.0, 4.0, 1.0, 0.0, 1.0, 3.0;
+
+  Matrix<double, Dynamic, Dynamic> LS = sigma.llt().matrixL();
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp:24]: This can just be `MatrixXd` or `auto`.

File: test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  EXPECT_NO_THROW(inv_wishart_cholesky_rng(3.0, LS, rng));
+  EXPECT_THROW(inv_wishart_cholesky_rng(2, LS, rng), std::domain_error);
+  EXPECT_THROW(inv_wishart_cholesky_rng(-1, LS, rng), std::domain_error);
+  LS(2, 2) = -1.0;
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp:29]: You don't need the `.0` everywhere.  The compiler figures out whether constants should be `double` or `int`.  The only exception is arithmetic with them. 

Now for the tests like the following, it makes sense to try both integer and floating point args to make sure the polymorphism is handled correctly.

File: test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+  Sigma << 1, 0, 0, 1;
+  Sigma_non_pos_def << -1, 0, 0, 1;
+
+  unsigned int dof = 5;
+
+  EXPECT_NO_THROW(inv_wishart_cholesky_rng(dof, Sigma, rng));
+  EXPECT_THROW(inv_wishart_cholesky_rng(dof, Sigma_non_pos_def, rng),
+               std::domain_error);
+}
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp:50]: I'd like to see tests of boundary conditions like size 0 and size 1 inputs.

File: test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  using stan::math::inv_wishart_cholesky_rng;
+  using stan::test::unit::expect_symmetric;
+  using stan::test::unit::spd_rng;
+
+  boost::random::mt19937 rng;
+
+  for (int k = 1; k < 20; ++k)
+    for (double nu = k - 0.9; nu < k + 10; ++nu)
+      for (int n = 0; n < 10; ++n)
+        expect_symmetric(stan::math::multiply_lower_tri_self_transpose(
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp:64]: Isn't this guaranteed to hold because it's the product of a matrix and its transpose?  Or to ask this another way, what can go wrong here?

File: test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  Sigma.resize(3, 3);
+  Sigma.setIdentity();
+  Y.resize(3, 3);
+  Y.setIdentity();
+  nu = 3;
+  EXPECT_NO_THROW(inv_wishart_cholesky_lpdf(Y, nu, Sigma));
+  nu = 2;
+  EXPECT_THROW(inv_wishart_cholesky_lpdf(Y, nu, Sigma), std::domain_error);
+}
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp:146]: This should also test that it does the right thing with 0 x 0 inputs (just throw an exception or return 0).

File: test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  Sigma.resize(2, 2);
+  Y.resize(2, 1);
+  EXPECT_THROW(inv_wishart_cholesky_lpdf(Y, nu, Sigma), std::domain_error);
+
+  nu = 5;
+  Sigma.resize(2, 2);
+  Y.resize(3, 3);
+  EXPECT_THROW(inv_wishart_cholesky_lpdf(Y, nu, Sigma), std::invalid_argument);
+
+  Sigma.resize(3, 3);
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp:138]: [optional]

I really don't like this resizing pattern because I can't easily tell what `Sigma` is after line 134, etc.  Is this getting rejected because it's not a proper Cholesky factor, or because `Sigma` and `Y` are different sizes?

I'd find these cleaner if line 136 was just:

```cpp
EXPECT_THROW(inv_wishart_cholesky_lpdf(MatrixXd::Identity(3, 3), 5, MatrixXd::Identity(2, 2));
```

It's very clear here that the throw is because the two args are different sizes.

If you really want to define the intermediate, I'd suggest conventional naming for identity matrices:

```cpp
auto I3 = MatrixXd::Identity(3, 3);
test(...I3...)
```

File: test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+  EXPECT_FLOAT_EQ(0.0,
+                  stan::math::inv_wishart_cholesky_lpdf<true>(LY, dof, LS));
+}
+TEST(ProbDistributionsInvWishartCholesky, Error) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+
+  using stan::math::inv_wishart_cholesky_lpdf;
+  Matrix<double, Dynamic, Dynamic> Sigma;
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp:113]: All of the `Matrix<double, Dynamic, Dynamic>` should be replaced with `MatrixXd`.

File: test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      1.091046, 0.8530256, 2.5786609;
+
+  Matrix<double, Dynamic, Dynamic> Sigma(3, 3);
+  Sigma << 7.785215, 3.059788, 1.107166, 3.059788, 10.3515035, -0.1232598,
+      1.107166, -0.1232598, 7.7623386;
+
+  Matrix<double, Dynamic, Dynamic> LY = Y.llt().matrixL();
+  Matrix<double, Dynamic, Dynamic> LS = Sigma.llt().matrixL();
+
+  double dof = 4.0;
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/inv_wishart_cholesky_test.cpp:103]: Don't need all the `.0` everywhere, as in 103 or 105.

**Review by bob-carpenter — CHANGES_REQUESTED** [review]
Mostly comments like for the first review, so hopefully just as easy to fix.

The biggest issue is that we want autodiff tests for anything that's going to be used with autodiff in Stan.

File: stan/math/prim/prob/inv_wishart_cholesky_rng.hpp [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+                                                const Eigen::MatrixXd& L,
+                                                RNG& rng) {
+  using Eigen::MatrixXd;
+  static const char* function = "inv_wishart_cholesky_rng";
+  index_type_t<MatrixXd> k = L.rows();
+  check_greater(function, "degrees of freedom > dims - 1", nu, k - 1);
+  check_cholesky_factor(function, "scale parameter", L);
+
+  MatrixXd L_inv = mdivide_left_tri<Eigen::Lower>(L);
+  return mdivide_left_tri<Eigen::Lower>(wishart_cholesky_rng(nu, L_inv, rng));
```
[the remark this answers is not in the document]
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_rng.hpp:23]: I don't know. It follows the convention from the `inv_wishart_rng`. Maybe @bgoodri knows?

File: stan/math/prim/prob/inv_wishart_cholesky_rng.hpp [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/prob/wishart_cholesky_rng.hpp>
+#include <stan/math/prim/prob/wishart_rng.hpp>
+
+namespace stan {
+namespace math {
+
+template <class RNG>
```
[the remark this answers is not in the document]
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_rng.hpp:12]: done

File: stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    lp += k * LOG_TWO * (1 - 0.5 * nu_ref);
+  }
+
+  if (include_summand<propto, T_dof>::value) {
+    lp -= lmgamma(k, 0.5 * nu_ref);
+  }
+
+  if (include_summand<propto, T_dof, T_scale, T_y>::value) {
+    auto LYinvLS = mdivide_left_tri<Eigen::Lower>(LY_ref, LS_ref);
+
```
[the remark this answers is not in the document]
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:74]: done

File: stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+/** \ingroup multivar_dists
+ * The log of the Inverse Wishart density for the given Cholesky factor, degrees
+ of freedom,
+ * and scale Cholesky factor matrix.
+ *
+ * The scale matrix, LS, must be a lower Cholesky factor.
+ * Dimension, k, is implicit.
+ * nu must be greater than k-1
+ *
+ * The change of variables from the input positive-definite matrix to
```
[the remark this answers is not in the document]
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:23]: done

File: stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp [hunk 13]
```diff
@@ -12,8 +12,8 @@ namespace stan {
 namespace math {
 
 /** \ingroup multivar_dists
- * The log of the Inverse Wishart density for the given Cholesky factor, degrees of
- freedom,
+ * The log of the Inverse Wishart density for the given Cholesky factor, degrees
```
[the remark this answers is not in the document]
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:15]: done

File: test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp (hunk 1 above)
Comment (spinkney) [reply test/unit/math/prim/prob/inv_wishart_cholesky_rng_test.cpp:24]: done

File: stan/math/prim/prob/inv_wishart_cholesky_rng.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply stan/math/prim/prob/inv_wishart_cholesky_rng.hpp]: We can leave this if we don't hear back from @bgoodri.  This has some pointers of what they're doing in Julia to try to catch up with R's implementation:  https://discourse.julialang.org/t/julia-vs-r-speed-for-sampling-the-wishart-distribution/21352/15

[the remark this answers is not in the document]
Comment (bgoodri) [reply stan/math/prim/prob/inv_wishart_cholesky_rng.hpp]: We have `chol2inv()` but I don't know if it would help all that much.

**Comment by spinkney** [comment]
@bob-carpenter this basically follows the `wishart_cholesky` code. I updated with the comments from that PR. This is ready for review.

**Comment by spinkney** [comment]
@bob-carpenter is there a chance you can review this week? I'm going to be out for a couple weeks starting next week and I'd like this to get in prior to the feature freeze.

**Comment by bob-carpenter** [comment]
Now that I know more about what I don't know, I'm going to have to opt out of trying to write or review C++.  I don't know who does understand this well enough, but I'd try @rok-cesnovar or @SteveBronder or @andrjohns.

**Comment by spinkney** [comment]
Got it. @SteveBronder then?

File: stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+  check_size_match(function, "Rows of random variable", L_S.rows(),
+                   "columns of random variable", L_S.cols());
+
+  T_L_Y_ref L_Y_ref = L_Y;
+  T_nu_ref nu_ref = nu;
+  T_L_S_ref L_S_ref = L_S;
+
+  check_greater(function, "Degrees of freedom parameter", nu_ref, k - 1);
+  check_positive(function, "Cholesky Random variable", L_Y_ref.diagonal());
+  check_positive(function, "columns of Cholesky Random variable",
```
Comment (rok-cesnovar) [inline stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:66]: Random is sometimes capitalized but not always in these check calls. Let's make this consistent. It seems to me that we use the capitalized version in other checks across math?

File: stan/math/prim/prob/wishart_cholesky_lpdf.hpp [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+#include <stan/math/prim/fun/constants.hpp>
+
+namespace stan {
+namespace math {
+/** \ingroup multivar_dists
+ * Return the natural logarithm of the unnormalized Wishart density of the
+ specified
+ * lower-triangular Cholesky factor variate, positive degrees of freedom, and
+ lower-triangular
+ * Cholesky factor of the scale matrix.
```
Comment (rok-cesnovar) [inline stan/math/prim/prob/wishart_cholesky_lpdf.hpp:19]: Same comment as above.

File: stan/math/prim/prob/wishart_cholesky_lpdf.hpp [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  check_size_match(function, "Rows of random variable", L_Y.rows(),
+                   "columns of scale parameter", L_S.rows());
+
+  T_L_Y_ref L_Y_ref = L_Y;
+  T_nu_ref nu_ref = nu;
+  T_L_S_ref L_S_ref = L_S;
+
+  check_greater(function, "Degrees of freedom parameter", nu_ref, k - 1);
+  check_positive(function, "Cholesky random variable", L_Y_ref.diagonal());
+  check_positive(function, "columns of Cholesky Random variable",
```
Comment (rok-cesnovar) [inline stan/math/prim/prob/wishart_cholesky_lpdf.hpp:62]: Same as above with capitalization.

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
The C++ looks great - just some minor comment stuff to clean up and error message consistency.

File: stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp (hunk 14 above)
Comment (spinkney) [reply stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:66]: so I shouldn't change anything?

Comment (rok-cesnovar) [reply stan/math/prim/prob/inv_wishart_cholesky_lpdf.hpp:65]: It's not capitalized in lines 53-58.

**Comment by spinkney** [comment]
@rok-cesnovar expression tests passed! Pr is in stanc3 and docs

**Review by rok-cesnovar — APPROVED** [review]
